url587829ef81_champdavis_KhovanovHomology_tgz-pJ8-KhovanovMapJ8/src/KhovanovMap.java

```diff
@@ -0,0 +1,183 @@
+import org.apache.commons.lang3.StringUtils;
+import java.util.*;
+
+public class KhovanovMap{
+    
+    private Knot theKnot;  //root knot
+    private Knot[] resolvedKnots;
+    private KhovanovRow[] rows; 
+    private Stack<Map> maps;
+
+    
+    public KhovanovMap(Knot kn){
+        this.theKnot = kn;
+        this.rows = new KhovanovRow[this.theKnot.numCrossings()+1];
+        
+        int n = this.rows.length-1; //COMBINATION FINDER since length of each row is based on combinations....it's nCk in this case
+        int nCk = 1;
+        for (int k = 0; k <= n; k++) {
+            this.rows[k] = new KhovanovRow(nCk);
+            nCk = nCk * (n-k) / (k+1);
+        }
+            
+        //this.resolvedKnots = new Knot[(int) Math.pow(2,this.theKnot.numCrossings())];
+        
+        fillResolvedKnots(this.theKnot);
+        setMaps(); 
+            
+    }
+    
+    public void fillResolvedKnots(Knot k){
+        if(k.getLeftKnot()==null){
+            int numOnes = StringUtils.countMatches(k.getBinCode(), "1");
+            this.rows[numOnes].add(new KhovanovMapNode(k.getBinCode(),k.getCircles()));
+            //this.resolvedKnots[Integer.parseInt(k.getBinCode(), 2)] = k; //puts the knot in its right place in the array based on its binary code
+        }
+            
+
+        else{
+            fillResolvedKnots(k.getLeftKnot());
+            fillResolvedKnots(k.getRightKnot());
+        }
+    }
+    
+    public void setMaps(){       
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.